### PR TITLE
ocamlPackages.mirage-crypto: 0.10.7 -> 0.11.0

### PR DESCRIPTION
**Pull request by Zimmi48**
###### Description of changes

Initially, the aim was just to update mirage-crypto to the latest version (which included a breakdown into a new package), but it turned out necessary to update many reverse dependencies and to package several new packages that were introduced during their last update. One package needed a patch that is not yet present in a release. One package (ocaml-git) is still not patched, so was marked as broken. I don't know what the process is in this case. It should not be too complicated to prepare a patch inspired by how the other packages were fixed (I have seen the same error in several other packages before updating them), but I have not investigated this, as I've already spent way longer than initially planned on this update.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

```
4 packages added:
ocamlPackages.dns-client-lwt (init at 7.0.1) ocamlPackages.dns-client-mirage (init at 7.0.1) ocamlPackages.mirage-crypto-rng-lwt (init at 0.11.0) ocamlPackages.tls-lwt (init at 0.16.0)

81 packages updated:
eliom liquidsoap ocaml4.14.1-async_rpc_websocket ocamlPackages.awa (0.1.1 → 0.1.2) ocamlPackages.awa-lwt (0.1.1 → 0.1.2) ocamlPackages.awa-mirage (0.1.1 → 0.1.2) ocaml4.14.1-bonsai ocaml4.14.1-ca-certs ocaml4.14.1-ca-certs-nss ocaml4.14.1-chacha ocaml4.14.1-cohttp-async ocaml4.14.1-cohttp-lwt-unix ocaml4.14.1-cohttp-mirage ocaml4.14.1-cohttp_async_websocket ocaml4.14.1-comby ocamlPackages.conduit (6.1.0 → 6.2.0) ocamlPackages.conduit-async (6.1.0 → 6.2.0) ocamlPackages.conduit-lwt (6.1.0 → 6.2.0) ocamlPackages.conduit-lwt-unix (6.1.0 → 6.2.0) ocamlPackages.conduit-mirage (6.1.0 → 6.2.0) ocaml4.14.1-curly ocamlPackages.dns (6.4.1 → 7.0.1) ocamlPackages.dns-certify (6.4.1 → 7.0.1) ocamlPackages.dns-cli (6.4.1 → 7.0.1) ocamlPackages.dns-client (6.4.1 → 7.0.1) ocamlPackages.dns-mirage (6.4.1 → 7.0.1) ocamlPackages.dns-resolver (6.4.1 → 7.0.1) ocamlPackages.dns-server (6.4.1 → 7.0.1) ocamlPackages.dns-stub (6.4.1 → 7.0.1) ocamlPackages.dns-tsig (6.4.1 → 7.0.1) ocamlPackages.dnssec (6.4.1 → 7.0.1) ocaml4.14.1-dune-release ocaml4.14.1-erm_xmpp ocaml4.14.1-git ocaml4.14.1-git-mirage ocaml4.14.1-git-paf ocaml4.14.1-git-unix ocaml4.14.1-graphql-cohttp ocamlPackages.happy-eyeballs (0.4.0 → 0.5.0) ocamlPackages.happy-eyeballs-lwt (0.4.0 → 0.5.0) ocamlPackages.happy-eyeballs-mirage (0.4.0 → 0.5.0) ocaml4.14.1-hkdf ocaml4.14.1-irmin-git ocaml4.14.1-irmin-graphql ocaml4.14.1-irmin-http ocaml4.14.1-irmin-mirage-git ocaml4.14.1-irmin-mirage-graphql jackline (2022-05-27 → 2023-02-24) ocamlPackages.letsencrypt (0.4.1 → 0.5.0) ocamlPackages.letsencrypt-app (0.4.1 → 0.5.0) ocamlPackages.letsencrypt-dns (0.4.1 → 0.5.0) ocaml4.14.1-ligo ocaml4.14.1-mimic-happy-eyeballs ocamlPackages.mirage-crypto (0.10.7 → 0.11.0) ocamlPackages.mirage-crypto-ec (0.10.7 → 0.11.0) ocamlPackages.mirage-crypto-pk (0.10.7 → 0.11.0) ocamlPackages.mirage-crypto-rng (0.10.7 → 0.11.0) ocamlPackages.mirage-crypto-rng-async (0.10.7 → 0.11.0) ocamlPackages.mirage-crypto-rng-mirage (0.10.7 → 0.11.0) ocaml4.14.1-mrmime ocaml4.14.1-ocsigen-start ocaml4.14.1-ocsigen-toolkit ocaml4.14.1-ocsigenserver ocaml4.14.1-ocsipersist ocaml4.14.1-ocsipersist-pgsql ocaml4.14.1-ocsipersist-sqlite ocaml4.14.1-opium ocaml4.14.1-otr ocaml4.14.1-paf ocaml4.14.1-paf-cohttp ocaml4.14.1-paf-le ocaml4.14.1-pbkdf ocaml4.14.1-piaf ocaml4.14.1-plotkicadsch ocaml4.14.1-resto-cohttp-self-serving-client ocaml4.14.1-resto-cohttp-server ocaml4.14.1-telegraml-unstable ocamlPackages.tls (0.15.4 → 0.16.0) ocamlPackages.tls-async (0.15.4 → 0.16.0) ocamlPackages.tls-mirage (0.15.4 → 0.16.0) ocamlPackages.x509 (0.16.2 → 0.16.4)

$ nix --experimental-features nix-command build --no-link --keep-going --no-allow-import-from-derivation --option build-use-sandbox relaxed -f /home/theo/.cache/nixpkgs-review/rev-f61f54ce3f0bf6f086a2586f3ba85923bbc83c8e/build.nix
10 packages marked as broken and skipped:
ocamlPackages.git ocamlPackages.git-mirage ocamlPackages.git-paf ocamlPackages.git-unix ocamlPackages.irmin-git ocamlPackages.irmin-graphql ocamlPackages.irmin-http ocamlPackages.irmin-mirage-git ocamlPackages.irmin-mirage-graphql ocamlPackages.plotkicadsch

75 packages built:
comby dune-release jackline ligo liquidsoap ocamlPackages.async_rpc_websocket ocamlPackages.awa ocamlPackages.awa-lwt ocamlPackages.awa-mirage ocamlPackages.bonsai ocamlPackages.ca-certs ocamlPackages.ca-certs-nss ocamlPackages.chacha ocamlPackages.cohttp-async ocamlPackages.cohttp-lwt-unix ocamlPackages.cohttp-mirage ocamlPackages.cohttp_async_websocket ocamlPackages.conduit ocamlPackages.conduit-async ocamlPackages.conduit-lwt ocamlPackages.conduit-lwt-unix ocamlPackages.conduit-mirage ocamlPackages.curly ocamlPackages.dns ocamlPackages.dns-certify ocamlPackages.dns-cli ocamlPackages.dns-client ocamlPackages.dns-client-lwt ocamlPackages.dns-client-mirage ocamlPackages.dns-mirage ocamlPackages.dns-resolver ocamlPackages.dns-server ocamlPackages.dns-stub ocamlPackages.dns-tsig ocamlPackages.dnssec ocamlPackages.eliom ocamlPackages.erm_xmpp ocamlPackages.graphql-cohttp ocamlPackages.happy-eyeballs ocamlPackages.happy-eyeballs-lwt ocamlPackages.happy-eyeballs-mirage ocamlPackages.hkdf ocamlPackages.letsencrypt ocamlPackages.letsencrypt-app ocamlPackages.letsencrypt-dns ocamlPackages.mimic-happy-eyeballs ocamlPackages.mirage-crypto ocamlPackages.mirage-crypto-ec ocamlPackages.mirage-crypto-pk ocamlPackages.mirage-crypto-rng ocamlPackages.mirage-crypto-rng-async ocamlPackages.mirage-crypto-rng-lwt ocamlPackages.mirage-crypto-rng-mirage ocamlPackages.mrmime ocamlPackages.ocsigen-start ocamlPackages.ocsigen-toolkit ocamlPackages.ocsigen_server ocamlPackages.ocsipersist ocamlPackages.ocsipersist-pgsql ocamlPackages.ocsipersist-sqlite ocamlPackages.opium ocamlPackages.otr ocamlPackages.paf ocamlPackages.paf-cohttp ocamlPackages.paf-le ocamlPackages.pbkdf ocamlPackages.piaf ocamlPackages.resto-cohttp-self-serving-client ocamlPackages.resto-cohttp-server ocamlPackages.telegraml ocamlPackages.tls ocamlPackages.tls-async ocamlPackages.tls-lwt ocamlPackages.tls-mirage ocamlPackages.x509
```

### CODE REVIEW

**Comment by vbgl** [comment]
The list of packages that are broken by this PR seems a bit long. It might be a good idea to wait for the availability of the patch for the `git` library.

**Comment by Zimmi48** [comment]
I've prepared a patch for ocaml-git (PR opened at https://github.com/mirage/ocaml-git/pull/612) and updated the PR to include it. There are no longer any broken packages. Results of `nix-review`:

>4 packages added:
ocamlPackages.dns-client-lwt (init at 7.0.1) ocamlPackages.dns-client-mirage (init at 7.0.1) ocamlPackages.mirage-crypto-rng-lwt (init at 0.11.0) ocamlPackages.tls-lwt (init at 0.16.0)
>
>81 packages updated:
eliom liquidsoap ocaml4.14.1-async_rpc_websocket ocamlPackages.awa (0.1.1 → 0.1.2) ocamlPackages.awa-lwt (0.1.1 → 0.1.2) ocamlPackages.awa-mirage (0.1.1 → 0.1.2) ocaml4.14.1-bonsai ocaml4.14.1-ca-certs ocaml4.14.1-ca-certs-nss ocaml4.14.1-chacha ocaml4.14.1-cohttp-async ocaml4.14.1-cohttp-lwt-unix ocaml4.14.1-cohttp-mirage ocaml4.14.1-cohttp_async_websocket ocaml4.14.1-comby ocamlPackages.conduit (6.1.0 → 6.2.0) ocamlPackages.conduit-async (6.1.0 → 6.2.0) ocamlPackages.conduit-lwt (6.1.0 → 6.2.0) ocamlPackages.conduit-lwt-unix (6.1.0 → 6.2.0) ocamlPackages.conduit-mirage (6.1.0 → 6.2.0) ocaml4.14.1-curly ocamlPackages.dns (6.4.1 → 7.0.1) ocamlPackages.dns-certify (6.4.1 → 7.0.1) ocamlPackages.dns-cli (6.4.1 → 7.0.1) ocamlPackages.dns-client (6.4.1 → 7.0.1) ocamlPackages.dns-mirage (6.4.1 → 7.0.1) ocamlPackages.dns-resolver (6.4.1 → 7.0.1) ocamlPackages.dns-server (6.4.1 → 7.0.1) ocamlPackages.dns-stub (6.4.1 → 7.0.1) ocamlPackages.dns-tsig (6.4.1 → 7.0.1) ocamlPackages.dnssec (6.4.1 → 7.0.1) ocaml4.14.1-dune-release ocaml4.14.1-erm_xmpp ocaml4.14.1-git ocaml4.14.1-git-mirage ocaml4.14.1-git-paf ocaml4.14.1-git-unix ocaml4.14.1-graphql-cohttp ocamlPackages.happy-eyeballs (0.4.0 → 0.5.0) ocamlPackages.happy-eyeballs-lwt (0.4.0 → 0.5.0) ocamlPackages.happy-eyeballs-mirage (0.4.0 → 0.5.0) ocaml4.14.1-hkdf ocaml4.14.1-irmin-git ocaml4.14.1-irmin-graphql ocaml4.14.1-irmin-http ocaml4.14.1-irmin-mirage-git ocaml4.14.1-irmin-mirage-graphql jackline (2022-05-27 → 2023-02-24) ocamlPackages.letsencrypt (0.4.1 → 0.5.0) ocamlPackages.letsencrypt-app (0.4.1 → 0.5.0) ocamlPackages.letsencrypt-dns (0.4.1 → 0.5.0) ocaml4.14.1-ligo ocaml4.14.1-mimic-happy-eyeballs ocamlPackages.mirage-crypto (0.10.7 → 0.11.0) ocamlPackages.mirage-crypto-ec (0.10.7 → 0.11.0) ocamlPackages.mirage-crypto-pk (0.10.7 → 0.11.0) ocamlPackages.mirage-crypto-rng (0.10.7 → 0.11.0) ocamlPackages.mirage-crypto-rng-async (0.10.7 → 0.11.0) ocamlPackages.mirage-crypto-rng-mirage (0.10.7 → 0.11.0) ocaml4.14.1-mrmime ocaml4.14.1-ocsigen-start ocaml4.14.1-ocsigen-toolkit ocaml4.14.1-ocsigenserver ocaml4.14.1-ocsipersist ocaml4.14.1-ocsipersist-pgsql ocaml4.14.1-ocsipersist-sqlite ocaml4.14.1-opium ocaml4.14.1-otr ocaml4.14.1-paf ocaml4.14.1-paf-cohttp ocaml4.14.1-paf-le ocaml4.14.1-pbkdf ocaml4.14.1-piaf ocaml4.14.1-plotkicadsch ocaml4.14.1-resto-cohttp-self-serving-client ocaml4.14.1-resto-cohttp-server ocaml4.14.1-telegraml-unstable ocamlPackages.tls (0.15.4 → 0.16.0) ocamlPackages.tls-async (0.15.4 → 0.16.0) ocamlPackages.tls-mirage (0.15.4 → 0.16.0) ocamlPackages.x509 (0.16.2 → 0.16.4)
>
>`$ nix --experimental-features nix-command build --no-link --keep-going --no-allow-import-from-derivation --option build-use-sandbox relaxed -f /home/theo/.cache/nixpkgs-review/rev-203a31a414408266fd8651de2ad80656228fa5ed/build.nix`
85 packages built:
comby dune-release jackline ligo liquidsoap ocamlPackages.async_rpc_websocket ocamlPackages.awa ocamlPackages.awa-lwt ocamlPackages.awa-mirage ocamlPackages.bonsai ocamlPackages.ca-certs ocamlPackages.ca-certs-nss ocamlPackages.chacha ocamlPackages.cohttp-async ocamlPackages.cohttp-lwt-unix ocamlPackages.cohttp-mirage ocamlPackages.cohttp_async_websocket ocamlPackages.conduit ocamlPackages.conduit-async ocamlPackages.conduit-lwt ocamlPackages.conduit-lwt-unix ocamlPackages.conduit-mirage ocamlPackages.curly ocamlPackages.dns ocamlPackages.dns-certify ocamlPackages.dns-cli ocamlPackages.dns-client ocamlPackages.dns-client-lwt ocamlPackages.dns-client-mirage ocamlPackages.dns-mirage ocamlPackages.dns-resolver ocamlPackages.dns-server ocamlPackages.dns-stub ocamlPackages.dns-tsig ocamlPackages.dnssec ocamlPackages.eliom ocamlPackages.erm_xmpp ocamlPackages.git ocamlPackages.git-mirage ocamlPackages.git-paf ocamlPackages.git-unix ocamlPackages.graphql-cohttp ocamlPackages.happy-eyeballs ocamlPackages.happy-eyeballs-lwt ocamlPackages.happy-eyeballs-mirage ocamlPackages.hkdf ocamlPackages.irmin-git ocamlPackages.irmin-graphql ocamlPackages.irmin-http ocamlPackages.irmin-mirage-git ocamlPackages.irmin-mirage-graphql ocamlPackages.letsencrypt ocamlPackages.letsencrypt-app ocamlPackages.letsencrypt-dns ocamlPackages.mimic-happy-eyeballs ocamlPackages.mirage-crypto ocamlPackages.mirage-crypto-ec ocamlPackages.mirage-crypto-pk ocamlPackages.mirage-crypto-rng ocamlPackages.mirage-crypto-rng-async ocamlPackages.mirage-crypto-rng-lwt ocamlPackages.mirage-crypto-rng-mirage ocamlPackages.mrmime ocamlPackages.ocsigen-start ocamlPackages.ocsigen-toolkit ocamlPackages.ocsigen_server ocamlPackages.ocsipersist ocamlPackages.ocsipersist-pgsql ocamlPackages.ocsipersist-sqlite ocamlPackages.opium ocamlPackages.otr ocamlPackages.paf ocamlPackages.paf-cohttp ocamlPackages.paf-le ocamlPackages.pbkdf ocamlPackages.piaf ocamlPackages.plotkicadsch ocamlPackages.resto-cohttp-self-serving-client ocamlPackages.resto-cohttp-server ocamlPackages.telegraml ocamlPackages.tls ocamlPackages.tls-async ocamlPackages.tls-lwt ocamlPackages.tls-mirage ocamlPackages.x509

File: pkgs/development/ocaml-modules/dns/client-lwt.nix [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    ipaddr
+    lwt
+    ca-certs
+    ca-certs-nss
+    happy-eyeballs
+    tcpip
+    tls-lwt
+    mtime
+    mirage-crypto-rng
+  ];
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/dns/client-lwt.nix:27]:
```suggestion
  propagatedBuildInputs = [
    dns
    dns-client
    ipaddr
    lwt
    ca-certs
    happy-eyeballs
    tls-lwt
    mtime
    mirage-crypto-rng
  ];
```

File: pkgs/development/ocaml-modules/dns/client-mirage.nix [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  propagatedBuildInputs = [
+    dns-client
+    domain-name
+    ipaddr
+    lwt
+    mirage-random
+    mirage-time
+    mirage-clock
+    ca-certs
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/dns/client-mirage.nix:23]:
```suggestion
```

File: pkgs/development/ocaml-modules/mirage-crypto/rng-lwt.nix [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+  inherit (mirage-crypto) version src;
+
+  doCheck = true;
+
+  buildInputs = [ dune-configurator ];
+  propagatedBuildInputs = [ mirage-crypto mirage-crypto-rng duration logs mtime ocaml_lwt ];
+
+  strictDeps = true;
+
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/mirage-crypto/rng-lwt.nix:15]:
```suggestion
```

This is not needed (this is the default value, set by `buildDunePackage`).

File: pkgs/development/ocaml-modules/tls/lwt.nix [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+  inherit (tls) src meta version;
+
+  minimalOCamlVersion = "4.11";
+  duneVersion = "3";
+
+  doCheck = true;
+
+  propagatedBuildInputs = [
+    cmdliner
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/tls/lwt.nix:14]:
```suggestion
```

**Review by vbgl — COMMENTED** [review]
Thanks a lot. I’ve suggested a few changes.

**Comment by Zimmi48** [comment]
Since my patch for ocaml-git has been integrated upstream, I have updated the PR to use `fetchpatch` instead of bundling the patch locally.

**Comment by vbgl** [comment]
Thanks.

**Pull request merged**